### PR TITLE
Implement HTTP applicator endpoints and simple server.

### PR DESCRIPTION
**Pull request by anthonybishopric**
This change introduces many more endpoints implemented in the
HTTP label applicator as well as a simple server that clients
can use.

This change also adds the Gorilla multiplexer.

### CODE REVIEW

File: pkg/labels/batcher.go [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+		}
+	}
+	b.pending = []chan []Labeled{}
+	b.batchInProgress = false
+}
+
+func (b *TypeBatcher) Retrieve() ([]Labeled, error) {
+	b.createMux.Lock()
+	if !b.batchInProgress {
+		b.batchInProgress = true
```
Comment (mpuncel) [inline pkg/labels/batcher.go]: might be neater to set this inside of `handleBatch()`

File: pkg/labels/batcher.go [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+	allLabels, err := b.applicator.ListLabels(b.labelType)
+	<-after
+	b.createMux.Lock()
+	defer b.createMux.Unlock()
+	b.queriesPerBatch.Update(int64(len(b.pending)))
+	for _, ch := range b.pending {
+		if err != nil {
+			close(ch)
+		} else {
+			ch <- allLabels
```
Comment (mpuncel) [inline pkg/labels/batcher.go]: you should `close(ch)` here as well

File: pkg/labels/batcher.go [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+	if !b.batchInProgress {
+		b.batchInProgress = true
+		go b.handleBatch()
+	}
+	respCh := make(chan []Labeled)
+	b.pending = append(b.pending, respCh)
+	b.createMux.Unlock()
+	res, ok := <-respCh
+	if !ok {
+		return nil, fmt.Errorf("Could not retrieve results")
```
Comment (mpuncel) [inline pkg/labels/batcher.go]: if we put this into use we'll probably want a better error message here

File: pkg/labels/http_applicator.go [hunk 4]
```diff
@@ -13,18 +15,25 @@ import (
...
 var _ Applicator = &httpApplicator{}
 
-func NewHttpApplicator(client *http.Client, matchesEndpoint *url.URL) (*httpApplicator, error) {
+func NewHTTPApplicator(client *http.Client, matchesEndpoint *url.URL) (*httpApplicator, error) {
+	// take only the base
+	base, err := url.Parse("/")
+	if err != nil {
+		return nil, err
+	}
+	matchesEndpoint = matchesEndpoint.ResolveReference(base)
```
Comment (mpuncel) [inline pkg/labels/http_applicator.go]: this should happen after the `nil` check

File: pkg/labels/http_applicator.go [hunk 5]
```diff
@@ -41,48 +50,196 @@ func NewHttpApplicator(client *http.Client, matchesEndpoint *url.URL) (*httpAppl
...
+		return nil, err
 	}
 	req.Header.Add("Accept", "application/json")
 
 	resp, err := h.client.Do(req)
 	if err != nil {
+		return nil, err
+	}
+
+	if err != nil {
```
Comment (mpuncel) [inline pkg/labels/http_applicator.go]: delete

File: pkg/labels/label_http_server.go [hunk 6]
```diff
@@ -0,0 +1,233 @@
...
+	return &labelHTTPServer{applicator, NewBatcher(applicator, batchTime), batchTime > 0, logger}
+}
+
+func (l *labelHTTPServer) Handler() *mux.Router {
+	r := mux.NewRouter()
+	r.Methods("GET").Path("/select").HandlerFunc(l.Select)
+	r.Methods("GET").Path("/labels/{type}/{id}").HandlerFunc(l.GetLabels)
+	r.Methods("GET").Path("/labels/{type}").HandlerFunc(l.ListLabels)
+	r.Methods("POST").Path("/labels/{type}/{id}/{name}").HandlerFunc(l.SetLabel)
+	r.Methods("POST").Path("/labels/{type}/{id}").HandlerFunc(l.SetLabels)
```
Comment (mpuncel) [inline pkg/labels/label_http_server.go]: nit maybe this should be `PUT`?

File: pkg/labels/batcher.go (hunk 1 above)
Comment (anthonybishopric) [inline pkg/labels/batcher.go]: You can't safely set this flag in a separate goroutine

File: pkg/labels/batcher.go (hunk 3 above)
Comment (anthonybishopric) [inline pkg/labels/batcher.go]: Yeah, there needs to be error logging when we get an error up a few lines.

File: pkg/labels/label_http_server.go (hunk 6 above)
Comment (anthonybishopric) [inline pkg/labels/label_http_server.go]: I was being lazy because of `client.Post`.

Comment (anthonybishopric) [inline pkg/labels/label_http_server.go]: I'll change it, no broken windows!

**Comment by anthonybishopric** [comment]
@bwester that's a good point. Serializability is lost when using this service to query label data as written. It would definitely be solved by using the hold time option as a way to collect requests prior to actually executing the query. 

I see a fairly significant downside to doing this. The total time that requests will be batched for goes from `max(holdTime, labelRetrievalTime)` to just `holdTime`. I think in practice this will somewhat minimize the batch count per query and increase the overall latency of match results, since you now need to wait `holdTime + labelRetrievalTime`.

However, I think your suggested solution still lets us perform batching and doesn't compromise data quality, which is critical. To your point, replication controllers do not act safely if the labels they've written are not immediately readable:

* RC abc has a pod marked with its ID
* RC def needs to `addOne`, it marks that pod with its ID
* RC def then checks again which nodes it manages and decides it needs to `addOne` again.

I will alter the batch handling to match your description and we can tune hold times as needed. We don't need another Faustian deal for performance gains.

**Pull request merged**